Handle requests with body

MaggieVL opened the pull request.

Commit:
Switch from ElmDefinition to TypeRef

This is the best way to match all the needed types, we can't match all
ElmDefinition constructors.

Commit:
Add new endpoints for post requests to test body queries

Commit:
Generate [Req] in a separate function
I needed to export it in order to print the [Req] to inspect them.

Commit:
Support a request body in elm queries

Add the value for the body as a separate parameter and support json and
empty body.

googleson78 commented on src/ServantElm.hs:
```diff
@@ -98,7 +129,7 @@ endpointInfoToElmQuery requestInfo =
       mkTypeSignature requestInfo
 
-    args = hsep ["urlBase", "toMsg"]
+    args = hsep ["urlBase", "toMsg", "bodyValue"]
```
The `"bodyValue"` constant is used in two places and it's necessary for it to be the same in both places - you should extract it as a constant.

MaggieVL replied:
done, did the same to the other argument values

googleson78 commented on src/ServantElm.hs:
```diff
@@ -126,12 +157,44 @@ mkUrl segments =
+    toMsgType =
+      fmap mkMsgType $ request ^. reqReturnType
+      where
+        mkMsgType x = "(Result Http.Error" <+> parens (elmTypeRef x) <+> "-> msg)"
```
Use `parens` instead of manually putting parentheses.

MaggieVL replied:
done

googleson78 commented on src/ServantElm.hs:
```diff
@@ -166,6 +229,42 @@ mkRequest request =
+    body =
+      case request ^. reqBody of
+        Just _ ->
+          "Http.jsonBody bodyValue"
```
Use the to-be extracted `bodyValue` constant here.

MaggieVL replied:
done

googleson78 commented on src/ServantElm.hs:
```diff
@@ -11,15 +11,34 @@
 import Elm.Ast
   ( ElmDefinition,
+    ElmPrim
+      ( ElmBool,
+        ElmChar,
+        ElmFloat,
+        ElmInt,
+        ElmList,
+        ElmMaybe,
+        ElmNever,
+        ElmPair,
+        ElmResult,
+        ElmString,
+        ElmTime,
+        ElmTriple,
+        ElmUnit
+      ),
```
Use a `(..)` import here instead imo.

MaggieVL replied:
done

googleson78 commented on src/ServantElm.hs:
```diff
@@ -166,6 +229,42 @@ mkRequest request =
+          "Http.emptyBody"
+
+typeRefDecoder :: TypeRef -> Doc ann
+typeRefDecoder (RefCustom TypeName {..}) = "decode" <> pretty (T.takeWhile (/= ' ') unTypeName)
```
You're only using one single filed - `unTypeName`. You can either use a field pun (`TypeName{unTypeName}`) or even use `unTypeName`as an accessor (`typeRefDecoder (RefCustom typeName) = "decode" <> pretty (T.takeWhile (/= ' ') unTypeName typeName)` instead.

MaggieVL replied:
done

googleson78 commented on src/ServantElm.hs:
```diff
@@ -166,6 +229,42 @@ mkRequest request =
+    "maybe"
+      <+> typeRefDecoder t
+  ElmResult l r ->
+    parens "elmStreetDecodeEither"
+      <+> typeRefDecoder l
+      <+> typeRefDecoder r
```
This seems wrong, you're only putting the parentheses around `elmStreetDecodeEither`, not around the entire expression.

MaggieVL replied:
done

googleson78 commented on src/ServantElm.hs:
```diff
@@ -166,6 +229,42 @@ mkRequest request =
+      <+> typeRefDecoder l
+      <+> typeRefDecoder r
+  ElmPair a b ->
+    parens "elmStreetDecodePair"
+      <+> typeRefDecoder a
+      <+> typeRefDecoder b
```
Same comment as above regarding parens

MaggieVL replied:
done

googleson78 commented on src/ServantElm.hs:
```diff
@@ -166,6 +229,42 @@ mkRequest request =
+      <+> typeRefDecoder a
+      <+> typeRefDecoder b
+  ElmTriple a b c ->
+    parens "elmStreetDecodeTriple"
+      <+> typeRefDecoder a
+      <+> typeRefDecoder b
+      <+> typeRefDecoder c
```
Same comment as above regarding parens

MaggieVL replied:
done

googleson78 commented on src/ServantElm.hs:
```diff
@@ -166,6 +229,42 @@ mkRequest request =
+      <+> typeRefDecoder a
+      <+> typeRefDecoder b
+      <+> typeRefDecoder c
+  ElmList l -> parens "Json.Decode.list" <+> typeRefDecoder l
```
Same comment as above regarding parens

MaggieVL replied:
done

googleson78 commented on src/ServantElm.hs:
```diff
@@ -166,6 +229,42 @@ mkRequest request =
+typeRefDecoder :: TypeRef -> Doc ann
+typeRefDecoder (RefCustom TypeName {..}) = "decode" <> pretty (T.takeWhile (/= ' ') unTypeName)
+typeRefDecoder (RefPrim elmPrim) = case elmPrim of
+  ElmUnit -> "Json.Decode.map (always ()) (Json.Decode.list Json.Decode.string)"
```
I think this should have parens inserted around it, because it might end up being an argument to another decoder

MaggieVL replied:
done

googleson78 commented on src/ServantElm.hs:
```diff
@@ -166,6 +229,42 @@ mkRequest request =
+  ElmTime -> "Iso.decoder"
+  ElmMaybe t ->
+    "maybe"
+      <+> typeRefDecoder t
```
I think this should have parens inserted around it, because it might end up being an argument to another decoder

MaggieVL replied:
done

googleson78 commented on src/ServantElm.hs:
```diff
@@ -166,6 +229,42 @@ mkRequest request =
+typeRefDecoder (RefCustom TypeName {..}) = "decode" <> pretty (T.takeWhile (/= ' ') unTypeName)
+typeRefDecoder (RefPrim elmPrim) = case elmPrim of
+  ElmUnit -> "Json.Decode.map (always ()) (Json.Decode.list Json.Decode.string)"
+  ElmNever -> "Json.Decode.fail \"Never is not possible\""
```
I think this should have parens inserted around it, because it might end up being an argument to another decoder

googleson78 replied:
Missed this one.

MaggieVL replied:
done

googleson78 commented on examples/servant-elm-street-main/src/ExampleAPI.hs:
```diff
@@ -9,25 +9,28 @@
 import Servant.Foreign (HasForeignType (..))
-import ServantElm (elmForAPI)
+
+data LangElm
+
+instance Elm a => HasForeignType LangElm ElmDefinition a where
+  typeFor _ _ proxyA = toElmDefinition proxyA
```
I don't think this should be here - it's already present in the library.

MaggieVL replied:
done

googleson78 commented on src/ServantElm.hs:
```diff
@@ -166,6 +229,42 @@ mkRequest request =
+          "Http.emptyBody"
+
+typeRefDecoder :: TypeRef -> Doc ann
+typeRefDecoder (RefCustom TypeName {..}) = "decode" <> pretty (T.takeWhile (/= ' ') unTypeName)
```
Why do you need the `takeWhile`?

MaggieVL replied:
it's not needed it seems, but this line is copied directly from elm-street code so as discussed we will keep it in case it's really needed

Commit:
Extract argument constants

We have the same text written twice so we need to make sure it's the
same in both places.

Commit:
Fix mkMsgType parenthesis

Commit:
Fix decoder parenthesis

Put them around the whole decoder and avoid the case when they are put
around primitive values unnecessarily

Commit:
Add parens around fail function

Commit:
No need to list all constructors

Commit:
Remove redundant instance declaration

Commit:
Add new endpoints to test decoders and type parentheses

Commit:
Use only one record field

No need to import all of them if we use only one.

Commit:
Add missing argument to crossOgirin fn

Commit:
Add space between queries

Commit:
Make bodyArg evaluated first

Commit:
Ignore elm-stuff directory

Commit:
Rename servant-elm-street-main to backend

Commit:
Add server example fro users

Commit:
Add elm users project

Commit:
Move libraby code

Commit:
Rename Json.Decode
We don't need to write the whole name each time, it's better to make it
shorter.

Commit:
Generate elm module and encoders

We need a function to generate the whole module containgn elm functions
and imports in order to enable importing them.

Commit:
Use encoders

Commit:
Add cors policy and elm code generation

We first use elm-street to generate types, en/decoders, then
generate module with elm queries to the specified api and start the
server.

Commit:
Add generated elm modules

In case anyone needs to see them without compiling the program.

googleson78 commented on examples/backend/package.yaml:
```diff
@@ -9,15 +9,15 @@ copyright:           "Copyright 2022 Chaos Software OOD"
 
 ghc-options:
-- -Wall
 - -Werror
+- -Wall
```
Try to avoid unnecessary shuffling like this in the future.

MaggieVL replied:
okay

googleson78 commented on examples/backend/src/ExampleAPI.hs:
```diff
@@ -9,33 +10,41 @@
-  deriving (Generic, Elm)
+  deriving (Generic)
+  deriving anyclass (Elm)
 
-instance ToJSON User
+instance ToJSON User where toJSON = elmStreetToJson
+
+instance FromJSON User where parseJSON = elmStreetParseJson
```
You can use `DerivingVia` instead, [as shown in the tutorial](https://github.com/Holmusk/elm-street#getting-started)

MaggieVL replied:
done

googleson78 commented on src/ServantElm.hs:
```diff
@@ -151,13 +222,14 @@ mkRequest request =
               <> indent4Spaces expect,
             "body ="
-              <> indent4Spaces "Http.emptyBody",
+              <> indent4Spaces body,
```
Why are you using `indent4Spaces` here? It causes unnecessary spacing, as shown in the generated example. Use `<+>` instead, I think.

MaggieVL replied:
done

googleson78 commented on src/ServantElm.hs:
```diff
@@ -166,6 +238,118 @@ mkRequest request =
+        Nothing ->
+          "Http.emptyBody"
+
+typeRefDecoder :: TypeRef -> Doc ann
```
This is taken from `elm-street`, right? You should add a comment, so that we know what we should ask `elm-street` to expose from their api eventually.

MaggieVL replied:
okay, done

googleson78 commented on src/ServantElm.hs:
```diff
@@ -166,6 +238,118 @@ mkRequest request =
+        <+> typeRefDecoder c
+  ElmList l -> parens $ "JD.list" <+> typeRefDecoder l
+
+typeRefEncoder :: TypeRef -> Doc ann
```
This is taken from `elm-street`, right? You should add a comment, so that we know what we should ask `elm-street` to expose from their api eventually.

MaggieVL replied:
done

googleson78 commented on src/ServantElm.hs:
```diff
@@ -166,6 +238,118 @@ mkRequest request =
+      vsep $
+        map
+          ("import" <+>)
+          [ "Core.Generated.Decoder" <+> "exposing" <+> parens "..",
```
This is taken from `elm-street`, right? You should add a comment, so that we know what we should ask `elm-street` to expose from their api eventually.

MaggieVL replied:
I wrote this myself, I didn't take it from them, do you mean the way they write their imports (they expose everything, this is an idea I saw from them, yes)

MaggieVL replied:
Should I write any comment in this case?

googleson78 replied:
These module names shouldn't be hardcoded, they should come from the settings.

googleson78 commented on src/ServantElm.hs:
```diff
@@ -126,12 +165,44 @@ mkUrl segments =
+elmTypeParenDoc :: TypeRef -> Doc ann
+elmTypeParenDoc = parens . elmTypeRefToDoc
+
+elmPrimToDoc :: ElmPrim -> Doc ann
```
This is taken from elm-street, right? You should add a comment, so that we know what we should ask elm-street to expose from their api eventually.

MaggieVL replied:
done

googleson78 commented on examples/frontend/src/Main.elm:
```diff
@@ -0,0 +1,111 @@
+view : Model -> Html Msg
+view model =
+    ul []
+        [ li [] [ text model.getAlbertResult ]
+        , li [] [ text model.getUsersResult ]
+        , li [] [ text model.postSignUpResult ]
+        , li [] [ text model.postSthResult ]
```
You should add some text indicating which endpoint each thing comes from imo

MaggieVL replied:
done

Commit:
Use DerivingVia with elm-street types

Commit:
Remove unnecessary spacing in request

Commit:
Add reminder for elm-street used functions

Commit:
Match result to endpoint

Commit:
Use settings parameter fields

It is better to use the fields given from settings because elm-street
might change the names some day.

googleson78 commented on src/ServantElm.hs:
```diff
@@ -19,20 +19,20 @@ module ServantElm
 where
 
-import Data.List (intercalate)
+import Data.List as L (intercalate)
```
Please use `as List`. `L` isn't really that much shorter, but it's less clear.

Commit:
Change List import alias name
